internal/consensus/pbts_test.go

```diff
@@ -1,56 +1,299 @@
 package consensus
 
 import (
+	"bytes"
+	"context"
 	"testing"
 	"time"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
+	"github.com/tendermint/tendermint/abci/example/kvstore"
+	tmpubsub "github.com/tendermint/tendermint/libs/pubsub"
 	tmtimemocks "github.com/tendermint/tendermint/libs/time/mocks"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
 	"github.com/tendermint/tendermint/types"
 )
 
+type pbtsTestHarness struct {
+	s                 pbtsTestConfiguration
+	observedState     *State
+	observedValidator *validatorStub
+	otherValidators   []*validatorStub
+	validatorClock    *tmtimemocks.Source
+
+	chainID string
+
+	ensureProposalCh, roundCh, blockCh, ensureVoteCh <-chan tmpubsub.Message
+
+	currentHeight int64
+	currentRound  int32
+}
+
+type pbtsTestConfiguration struct {
+	timestampParams            types.TimestampParams
+	timeoutPropose             time.Duration
+	genesisTime                time.Time
+	height2ProposalDeliverTime time.Time
+	height2ProposedBlockTime   time.Time
+}
+
+func newPBTSTestHarness(t *testing.T, tc pbtsTestConfiguration) pbtsTestHarness {
+	const validators = 4
+	cfg := configSetup(t)
+	clock := new(tmtimemocks.Source)
+	cfg.Consensus.TimeoutPropose = tc.timeoutPropose
+	consensusParams := types.DefaultConsensusParams()
+	consensusParams.Timestamp = tc.timestampParams
+
+	state, privVals := makeGenesisState(cfg, genesisStateArgs{
+		Params:     consensusParams,
+		Time:       tc.genesisTime,
+		Validators: validators,
+	})
+	cs := newState(state, privVals[0], kvstore.NewApplication())
+	vss := make([]*validatorStub, validators)
+	for i := 0; i < validators; i++ {
+		vss[i] = newValidatorStub(privVals[i], int32(i))
+	}
+	incrementHeight(vss[1:]...)
+
+	for _, vs := range vss {
+		vs.clock = clock
+	}
+	pubKey, err := vss[0].PrivValidator.GetPubKey(context.Background())
+	assert.NoError(t, err)
+
+	return pbtsTestHarness{
+		s:                 tc,
+		observedValidator: vss[0],
+		observedState:     cs,
+		otherValidators:   vss[1:],
+		validatorClock:    clock,
+		currentHeight:     1,
+		chainID:           cfg.ChainID(),
+		roundCh:           subscribe(cs.eventBus, types.EventQueryNewRound),
+		ensureProposalCh:  subscribe(cs.eventBus, types.EventQueryCompleteProposal),
+		blockCh:           subscribe(cs.eventBus, types.EventQueryNewBlock),
+		ensureVoteCh:      subscribeToVoterBuffered(cs, pubKey.Address()),
+	}
+}
+
+func (p *pbtsTestHarness) genesisHeight(t *testing.T) {
+	p.validatorClock.On("Now").Return(p.s.height2ProposedBlockTime).Times(8)
```

Comment:
Is "8" significant or is it just a "more times than I expect to care about"?

Reply:
8 is the number of readings of this clock that should occur during a single height. Once for each vote being cast. (4 validators, each cast 1 prevote and 1 precommit).